Refactor, cleanup, reorganize (#2)
cfrg/draft-irtf-cfrg-hybrid-kems: draft-irtf-cfrg-hybrid-kems.md

- Key Derivation Function (KDF) {{kdf}}; and
- Nominal Diffie-Hellman Group {{group}}.

These dependencies are defined in the following subsections.

## Key encapsulation mechanisms {#kems}

Key encapsulation mechanisms (KEMs) are cryptographic schemes that consist of
four algorithms:

- `KeyGen() -> (pk, sk)`: A probabilistic key generation algorithm, which
  generates a public encapsulation key `pk` and a secret decapsulation key
  `sk`, each of which are byte strings.
- `DeriveKey(seed) -> (pk, sk)`: A deterministic algorithm, which takes as
  input a seed `seed` and generates a public encapsulation key `pk` and a
  secret decapsulation key `sk`, each of which are byte strings.
- `Encaps(pk) -> (ct, shared_secret)`: A probabilistic encapsulation
  algorithm, which takes as input a public encapsulation key `pk` and outputs
  a ciphertext `ct` and shared secret `shared_secret`.
- `Decaps(sk, ct) -> shared_secret`: A decapsulation algorithm, which takes
  as input a secret decapsulation key `sk` and ciphertext `ct` and outputs a
  shared secret `shared_secret`.

KEMs can also provide a deterministic version of `Encaps`, denoted `EncapsDerand`,
with the following signature:

> - `EncapsDerand(pk, eseed) -> (ct, shared_secret)`: A deterministic encapsulation
>    algorithm, which takes as input a public encapsulation key `pk` and seed
>    `eseed`, and outputs a ciphertext `ct` and shared secret `shared_secret`.

Reviewer: ```suggestion
- `EncapsDerand(pk, randomness) -> (ct, shared_secret)`: A deterministic encapsulation
   algorithm, which takes as input a public encapsulation key `pk` and randomness
   `randomness`, and outputs a ciphertext `ct` and shared secret `shared_secret`.
```

fit the whole preimage into the Keccak block input size, which would be nice
for performance. But that might be trying to hard to over-engineer this. -->

The rest of this section specifies the key generation, encapsulation, and decapsulation procedures for this hybrid KEM.

### Key generation

`QSF-SHA3-256-ML-KEM-768-P-256` KeyGen works as follows.

> <!-- TODO(caw): we need to wire-encode the keys before outputting them -->

Reviewer: 👍

  return sk, concat(pq_PK, trad_PK)
~~~

Similarly, `QSF-SHA3-256-ML-KEM-768-P-256` DeriveKey works as follows:

~~~
def DeriveKey(seed):
  (pq_SK, trad_SK, pq_PK, trad_PK) = expandDecapsulationKey(seed)
  return sk, concat(pq_PK, trad_PK)
~~~

### Encapsulation

Given an encapsulation key `pk`, `QSF-SHA3-256-ML-KEM-768-P-256` Encaps proceeds as follows.

~~~
def Encaps(pk):
  pq_PK = pk[0:1184]
  trad_PK = pk[1184:1217]
  (pq_SS, pq_CT) = ML-KEM-768.Encaps(pq_PK)
  ek = P-256.RandomScalar()
  trad_CT = P-256.ScalarBaseMult(ek)
  trad_SS = P-256.ScalarMult(trad_PK, ek)
  ss = SHA3-256(pq_SS, trad_SS, trad_CT, trad_PK, label)
  ct = concat(pq_CT, trad_CT)
  return (ss, ct)
~~~

`pk` is a 1217-byte encapsulation key resulting from KeyGen().

Encaps() returns the 32-byte shared secret `ss` and the 1121-byte ciphertext `ct`.

Note that `Encaps()` may raise an error if ML-KEM-768.Encaps fails, e.g., if it does not pass the check of {{FIPS203}} §7.2.

### Derandomized

For testing, it is convenient to have a deterministic version of encapsulation. In such
cases, an implementation can provide the following derandomized function.

~~~

> def EncapsDerand(pk, eseed):
>   pq_PK = pk[0:1184]
>   trad_PK = pk[1184:1217]
>   (pq_SS, pq_CT) = ML-KEM-768.EncapsDerand(pq_PK, eseed[0:32])
>   ek = eseed[32:65]
>   trad_CT = P-256.ScalarMultBase(ek)
>   trad_SS = P-256.ScalarMult(ek, trad_PK)
>   ss = SHA3-256(pq_SS, trad_SS, trad_CT, trad_PK, label)
>   ct = concat(pq_CT, trad_CT)

Reviewer: ```suggestion
def EncapsDerand(pk, randomness):
  pq_PK = pk[0:1184]
  trad_PK = pk[1184:1217]
  (pq_SS, pq_CT) = ML-KEM-768.EncapsDerand(pq_PK, randomness[0:32])
  ek = randomness[32:65]
  trad_CT = P-256.ScalarMultBase(ek)
  trad_SS = P-256.ScalarMult(ek, trad_PK)
  ss = SHA3-256(pq_SS, trad_SS, trad_CT, trad_PK, label)
  ct = concat(pq_CT, trad_CT)
```

  return (ss, ct)
~~~

> Note that `eseed` MUST be 65 bytes.

Reviewer: ```suggestion
Note that `randomness` MUST be 65 bytes.
```

  return sk, concat(pq_PK, trad_PK)
~~~

Similarly, `KitchenSink-HKDF-SHA-256-ML-KEM-768-X25519` DeriveKey works as follows:

~~~
def DeriveKey(seed):
  (pq_SK, trad_SK, pq_PK, trad_PK) = expandDecapsulationKey(seed)
  return sk, concat(pq_PK, trad_PK)
~~~

### Encapsulation

Given an encapsulation key `pk`, `KitchenSink-HKDF-SHA-256-ML-KEM-768-X25519` Encaps proceeds as follows.

~~~
def Encaps(pk):
  pq_PK = pk[0:1184]
  trad_PK = pk[1184:1216]
  (pq_SS, pq_CT) = ML-KEM-768.Encaps(pq_PK)
  ek = random(32)
  trad_CT = X25519(ek, 9)
  trad_SS = X25519(ek, trad_PK)
  ss = LabeledHKDF(pq_SS, trad_SS, pq_CT, pq_PK, trad_CT, trad_PK, label)
  ct = concat(pq_CT, trad_CT)
  return (ss, ct)
~~~

pk is a 1216-byte encapsulation key resulting from KeyGen().

Encaps() returns the 32-byte shared secret ss and the 1120-byte ciphertext ct.

Note that `Encaps()` may raise an error if ML-KEM-768.Encaps fails, e.g., if it does not pass the check of {{FIPS203}} §7.2.

### Derandomized

For testing, it is convenient to have a deterministic version of encapsulation. In such
cases, an implementation can provide the following derandomized function.

~~~

> def EncapsDerand(pk, eseed):
>   pq_PK = pk[0:1184]
>   trad_PK = pk[1184:1216]
>   (pq_SS, pq_CT) = PQ-KEM.EncapsDerand(pq_PK, eseed[0:32])
>   ek = eseed[32:64]

Reviewer: ```suggestion
def EncapsDerand(pk, randomness):
  pq_PK = pk[0:1184]
  trad_PK = pk[1184:1216]
  (pq_SS, pq_CT) = PQ-KEM.EncapsDerand(pq_PK, randomness[0:32])
  ek = randomness[32:64]
```

  trad_CT = X25519(ek, 9)
  trad_SS = X25519(ek, trad_PK)
  ss = LabeledHKDF(pq_SS, trad_SS, pq_CT, pq_PK, trad_CT, trad_PK, label)
  ct = concat(pq_CT, trad_CT)
  return (ss, ct)
~~~

> Note that `eseed` MUST be 64 bytes.

Reviewer: ```suggestion
Note that `randomness` MUST be 64 bytes.
```

  return sk, concat(pq_PK, trad_PK)
~~~

Similarly, `QSF-SHA3-256-ML-KEM-1024-P-384` DeriveKey works as follows:

~~~
def DeriveKey(seed):
  (pq_SK, trad_SK, pq_PK, trad_PK) = expandDecapsulationKey(seed)
  return sk, concat(pq_PK, trad_PK)
~~~

### Encapsulation

Given an encapsulation key `pk`, `QSF-SHA3-256-ML-KEM-1024-P-384` Encaps proceeds as follows.

~~~
def Encaps(pk):
  pq_PK = pk[0:1568]
  trad_PK = pk[1568:1629]
  (pq_SS, pq_CT) = ML-KEM-1024.Encaps(pq_PK)
  ek = P-384.RandomScalar()
  trad_CT = P-384.ScalarBaseMult(ek)
  trad_SS = P-384.ScalarMult(trad_PK, ek)
  ss = SHA3-256(pq_SS, trad_SS, trad_CT, trad_PK, label)
  ct = concat(pq_CT, trad_CT)
  return (ss, ct)
~~~

`pk` is a 1629-byte encapsulation key resulting from KeyGen().

Encaps() returns the 32-byte shared secret `ss` and the 1629-byte ciphertext `ct`.

Note that `Encaps()` may raise an error if ML-KEM-1024.Encaps fails, e.g., if it does not pass the check of {{FIPS203}} §7.2.

### Derandomized

For testing, it is convenient to have a deterministic version of encapsulation. In such
cases, an implementation can provide the following derandomized function.

~~~

> def EncapsDerand(pk, eseed):
>   pq_PK = pk[0:1568]
>   trad_PK = pk[1568:1629]
>   (pq_SS, pq_CT) = ML-KEM-1024.EncapsDerand(pq_PK, eseed[0:32])
>   ek = eseed[32:80]
>   trad_CT = P-384.ScalarMultBase(ek)
>   trad_SS = P-384.ScalarMult(ek, trad_PK)
>   ss = SHA3-256(pq_SS, trad_SS, trad_CT, trad_PK, label)
>   ct = concat(pq_CT, trad_CT)

Reviewer: ```suggestion
def EncapsDerand(pk, randomness):
  pq_PK = pk[0:1568]
  trad_PK = pk[1568:1629]
  (pq_SS, pq_CT) = ML-KEM-1024.EncapsDerand(pq_PK, randomness[0:32])
  ek = randomness[32:80]
  trad_CT = P-384.ScalarMultBase(ek)
  trad_SS = P-384.ScalarMult(ek, trad_PK)
  ss = SHA3-256(pq_SS, trad_SS, trad_CT, trad_PK, label)
  ct = concat(pq_CT, trad_CT)
```

  return (ss, ct)
~~~

> Note that `eseed` MUST be 80 bytes.

Reviewer: ```suggestion
Note that `randomness` MUST be 80 bytes.
```
